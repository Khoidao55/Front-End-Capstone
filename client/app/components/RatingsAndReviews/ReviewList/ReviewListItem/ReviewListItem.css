.reviewListItemStarRating {
  margin: 10px 0 0 10px;
  display: flex;
  flex-direction: row;
}

.review-username-date {
  margin-right: 10px;
  margin-left: auto;
}

.review-list-item {
  border-bottom: 1px solid grey;
}

.reviewListItemSummary {
  font-weight: bold;
  font-size: large;
  margin: 10px 0 0 10px;
}

.reviewListItemBody {
  margin: 10px 0 0 10px;
}

.review-show-more-button {
  background: none;
  border: none;
  padding: 0;
  font-family: arial, sans-serif;
  color: #069;
  text-decoration: underline;
  cursor: pointer;
}

.review-show-more-button-hidden {
  display: none;
}

.review-response-container {
  border: 1px solid #333;
  background-color: #dfdfdf;
  margin: 10px;
}

.review-response-header {
  margin: 10px 0 0 10px
}

.review-response-body {
  margin: 10px 0 10px 10px
}