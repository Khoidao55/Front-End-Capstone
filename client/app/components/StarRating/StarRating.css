#container {
  position: relative;
  width: 90px;
  height: 16px;
  background-color: white;
  display: flex;
  justify-content: flex-end;
}

.stars {
  position: absolute;
  display: flex;
  flex-direction: row;
}

.fa-star {
  font-size: 16px;
}

.fas {
  color: gold;
}

.empty {
  z-index: 3;
}

#cover {
  position: relative;
  width: 100%;
  height: 16px;
  background-color: white;
  z-index: 2;
}

#no-reviews {
  margin: 10px 0 0 10px
}