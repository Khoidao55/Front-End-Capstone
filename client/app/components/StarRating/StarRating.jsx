import React from 'react';

// eslint-disable-next-line react/prop-types
const StarReview = ({ reviewScore, setMargin = '20px 0 0 0' }) => {
  // Rounds the score to the nearest quarter
  const roundedScore = Math.round(reviewScore * 4) / 4;

  // Create a solely visual offset so a quarter/3-quarter
  // star *looks* more like a quarter than it actually does
  const testOffset = roundedScore - Math.round(roundedScore);
  let quarterOffset = testOffset === 0.25 || testOffset === -0.25 ? 2.5 : 0;
  // eslint-disable-next-line no-unused-expressions
  testOffset < 0 ? quarterOffset *= -1 : 0;

  // Turn it into a percantage of stars to show
  // This works in "reverse" since it's calculating the percentage
  // of *missing* stars rather than that of showing stars because
  // width extends/shrinks from the left
  const missingPercent = `${100 - roundedScore * 20 - quarterOffset}%`;

  return (
    <div id="container" style={{ margin: setMargin }}>
      <EmptyStars />
      <div id="cover" style={{ width: missingPercent }}> </div>
      <FilledStars />
    </div>
  );
};

const EmptyStars = () => (
  <div className="empty stars">
    <i className="far fa-star" />
    <i className="far fa-star" />
    <i className="far fa-star" />
    <i className="far fa-star" />
    <i className="far fa-star" />
  </div>
);

const FilledStars = () => (
  <div className="full stars">
    <i className="fas fa-star" />
    <i className="fas fa-star" />
    <i className="fas fa-star" />
    <i className="fas fa-star" />
    <i className="fas fa-star" />
  </div>
);

export default StarReview;
