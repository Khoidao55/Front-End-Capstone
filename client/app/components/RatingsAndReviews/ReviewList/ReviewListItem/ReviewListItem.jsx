/* eslint-disable react/prop-types */
import React, { useEffect, useState } from 'react';
import StarRating from '../../../StarRating/StarRating.jsx';
import ReviewListItemThumbnails from './ReviewListItemThumbnails/ReviewListItemThumbnails.jsx';
import './ReviewListItem.css';

const ReviewListItem = ({ review }) => {
  const [reviewTileBody, setReviewTileBody] = useState(review.body.substring(0, 250));
  const [reviewBodyButtonClass, setReviewBodyButtonClass] = useState('review-show-more-button');

  const monthArray = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
  const month = review.date.slice(5, 7);
  const day = review.date.slice(8, 10);
  const year = review.date.slice(0, 4);
  const formattedString = `${monthArray[month - 1]} ${day}, ${year}`;

  const handleShowMoreBody = () => {
    setReviewTileBody(review.body);
    setReviewBodyButtonClass('review-show-more-button-hidden');
  };

  const ReviewBodyRender = () => {
    if (review.body.length > 250) {
      return (
        <>
          <div className="reviewListItemBody">
            {reviewTileBody}
          </div>
          <div className="review-show-body-link">
            <button type="button" className={reviewBodyButtonClass} onClick={handleShowMoreBody}>show more</button>
          </div>
        </>
      );
    }
    return (
      <div className="reviewListItemBody">
        {reviewTileBody}
      </div>
    );
  };

  const ReviewItemThumbnailRender = () => {
    if (review.photos.length > 0) {
      return (
        <ReviewListItemThumbnails photos={review.photos} />
      );
    }
    return (<></>);
  };

  return (
    <div className="reviewListItem">
      <div className="reviewListItemStarRating">
        <StarRating reviewScore={review.rating} setMargin="0 0 0 0" />
        <div className="review-username-date">
          {review.reviewer_name}
          {', '}
          {formattedString}
        </div>
      </div>
      <div className="reviewListItemSummary">
        {review.summary}
      </div>
      <ReviewBodyRender />
      <ReviewItemThumbnailRender />
    </div>
  );
};


export default ReviewListItem;
