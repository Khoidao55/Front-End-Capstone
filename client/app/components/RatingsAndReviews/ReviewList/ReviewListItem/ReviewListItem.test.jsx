import React from 'react';
import { shallow } from 'enzyme';
import ReviewListItem from './ReviewListItem.jsx';
import StarRating from '../../../StarRating/StarRating.jsx';

let review = {
  "review_id": 111396,
  "rating": 4,
  "summary": "Doloribus aut rerum.",
  "recommend": true,
  "response": null,
  "body": "Voluptas dolor ea ut sed accusantium. Dicta magnam expedita. Ipsum reprehenderit qui blanditiis consectetur perspiciatis. Voluptatum commodi aut.",
  "date": "2020-03-23T00:00:00.000Z",
  "reviewer_name": "Tanner.Lemke",
  "helpfulness": 20,
  "photos": [
      {
          "id": 182303,
          "url": "https://images.unsplash.com/photo-1513531926349-466f15ec8cc7?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
      },
      {
          "id": 182304,
          "url": "https://images.unsplash.com/photo-1562542119-19d015b93c45?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2057&q=80"
      }
  ]
};

describe('review list item: ', () => {
  test('it renders when given a review', () => {
    const wrapper = shallow(<ReviewListItem review={review} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('it renders a star rating component: ', () => {
    const wrapper = shallow(<ReviewListItem review={review} />);

    expect(wrapper.find(StarRating).length).toBe(1);
  });

  test('it renders correct number of divs: ', () => {
    const wrapper = shallow(<ReviewListItem review={review} />);
    // TODO - update this test as I complete review list item
    expect(wrapper.find('div').length).toBe(4);
  });
});
