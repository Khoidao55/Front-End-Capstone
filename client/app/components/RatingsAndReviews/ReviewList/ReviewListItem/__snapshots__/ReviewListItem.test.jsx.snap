// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`review list item:  it renders when given a review 1`] = `ShallowWrapper {}`;
